### ISSUE TITLE
quick move from search list fails if first mail is already in target folder

### ISSUE THREAD

**Issue by FrankMittelbach** [RealRaven2000/QuickFolders#23]
If you want to move a selection of mails to a specific folder,  then that works well even if some of the mails are already in the target folder. However, it fails if the first of the messages is already in the target folder.

In my use case that is quite often the case: I select a whole thread and move it into a folder, but quite often the first of the mails (e.g., my initial message) is already in that folder and then it fails. Of course, one can select all and then unselect offending message, so there is a workaround, but it would be nice if the quick move would act on the others without the need to do that as it is easily overlooked first time and it works well if the messages are ordered differently.

**Comment by RealRaven2000** [comment]
> In my use case that is quite often the case: I select a whole thread and move it into a folder, but quite often the first of the mails (e.g., my initial message) is already in that folder and then it fails. 

how do you do this? Some sort of a unified / all messages view? I always work with my "real inbox" and then answer "in place" (from a folder) using the [copy sent to current](https://addons.thunderbird.net/en-US/thunderbird/addon/copy-sent-to-current/) Add-on.

>Of course, one can select all and then unselect offending message, so there is a workaround, but it would be nice if the quick move would act on the others without the need to do that

yeah it shouldn't really "crash" the move if it is already in place - can you provide a (text only please) debug log? Ctrl+Shift+J opens the log window, which you can clear before you do the move. then use the context menu to select all, copy to clipboard and dump it here in a comment please.

**Comment by FrankMittelbach** [comment]
> yeah it shouldn't really "crash" the move if one (or several) mails are already in place - can you provide a (text only please) debug log? Ctrl+Shift+J opens the log window, which you can clear before you do the move. then use the context menu to select all, copy to clipboard and dump it here in a comment please.

if to figure out first how the console works on a Mac with German locale, it looks quite different from you suggestions. If it is too complicated I'll try on a windows machine tonight (which also has the problem)

**Comment by RealRaven2000** [comment]
> have to figure out first how the console works on a Mac with German locale, it looks quite different from you suggestions. If it is too complicated I'll try on a windows machine tonight (which also has the problem)

Ok. The main difference is the keyboard shortcut - you can find it in the tools menu:
![image](https://user-images.githubusercontent.com/6314674/72358873-2c7cab80-36e5-11ea-9cd2-6c79632eb3db.png)

Tools > Developer Tools > Error Console

to log additional diagnostic data you can:

![image](https://user-images.githubusercontent.com/6314674/72359148-add43e00-36e5-11ea-8f6b-b907a06c6b4d.png)

- enable debug mode in QuickFolders Settings / Advanced Tab
- right-click Debug and toggle `extensions.quickfolders.debug.quickMove = true`

How do you display emails that are already moved in the source folder while moving to target? With a "normal" inbox, you shouldn't see them. Note: you can also "gather" mails to quickMove in multiple drag + drop operations (hence the counter) and do the quickMove for all of them later. I assumed it should be resilient to missing mails (at least move the remaining ones).

**Comment by FrankMittelbach** [comment]
> > have to figure out first how the console works on a Mac with German locale, it looks quite different from you suggestions. If it is too complicated I'll try on a windows machine tonight (which also has the problem)
> 
> Ok. The main difference is the keyboard shortcut - you can find it in the tools menu:

thanks for the detailed explanations. Actually the starting problem with a locale different to English is usually also to find the right "translations", for example "Tools" simply doesn't exist as menu you have to use "Extras" etc etc :-) in some cases guessing is easy but in others it is not. Anyway ...


> How do you display emails that are already moved in the source folder while moving to target? With a "normal" inbox, you shouldn't see them. Note: you can also "gather" mails to quickMove in multiple drag + drop operations (hence the counter) and do the quickMove for all of them later. I assumed it should be resilient to missing mails (at least move the remaining ones).

yes of course with the normal inbox view I don't see them. What I often do however is to select a mail and execute "In Themenbaum anzeigen" on the mac that's  shift-cmd-O which gives me an additional (virtual) tab with all related message by message id (I guess) and that from where I  use quickmove (see screenshot which show the tab which I used for the debugging data)

<img width="1434" alt="Screenshot 2020-01-15 at 10 10 22" src="https://user-images.githubusercontent.com/1062981/72420668-6ef2c680-377f-11ea-9271-3bc4316313c6.png">

I also often use virtual folders like "mails from the last 3 days in a defined set of folders" and execute it from there.

Anyway, here is the debug data. Basically I  used select all in the virtual table shown above and then quickmove choosing a target folder archive/2020 because the first was already in that folder:

```
QuickFolders 9:53:32.682  [51891 ms]	 
toggleMoveModeSearchBox(true)
QuickFolders 9:53:32.682  [0 ms]	 
toggleMoveModeSearchBox(true)
QuickFolders 9:53:32.682  [0 ms]	 
QuickFolders.quickMove.update()
isActive = true
QuickFolders 9:53:32.682  [0 ms]	 
toggleMoveModeSearchBox(true)
QuickFolders {QUICKMOVE} 9:53:32.682  [0 ms]	 
findFolder(true, quickMove)
QuickFolders 9:53:32.682  [0 ms]	 
toggleMoveModeSearchBox(true)
QuickFolders 9:53:37.615  [4933 ms]	 
findFolder (archiv/2020)
QuickFolders {QUICKMOVE} 9:53:37.629  [14 ms]	 
Enter key forces match
QuickFolders {QUICKMOVE} 9:53:37.655  [26 ms]	 
rememberLastFolder([xpconnect wrapped (nsISupports, nsIRDFResource, nsIMsgFolder, nsISupportsWeakReference)], archiv)
QuickFolders {QUICKMOVE} 9:53:37.655  [0 ms]	 
Storing: archiv/2020
QuickFolders {QUICKMOVE} 9:53:37.655  [0 ms]	 
quickMove.execute() , tabMode = glodaList
QuickFolders {QUICKMOVE} 9:53:37.655  [0 ms]	 
QuickFolders.Util.moveMessages: target = 2020, makeCopy=false
QuickFolders 9:53:37.656  [1 ms]	 
slideAlert: Nothing to do: Message is already in folder: 2020
QuickFolders {QUICKMOVE} 9:53:37.656  [0 ms]	 
After move Messages actionCount: 0 resetting menu
QuickFolders 9:53:37.656  [0 ms]	 
QuickFolders.quickMove.update()
isActive = false
QuickFolders 9:53:37.656  [0 ms]	 
toggleMoveModeSearchBox(false)
QuickFolders {QUICKMOVE} 9:53:37.656  [0 ms]	 
findFolder(false, null)
QuickFolders 9:53:37.677  [21 ms]	 
QuickFolders.quickMove.update()
isActive = false
QuickFolders 9:53:37.677  [0 ms]	 
toggleMoveModeSearchBox(false)
QuickFolders {QUICKMOVE} 9:53:37.678  [1 ms]	 
findFolder(false, null)
QuickFolders 9:53:37.679  [1 ms]	 
util.touch(2020)
1579078417
15/1 9:53:37
QuickFolders {QUICKMOVE} 9:53:37.679  [0 ms]	 
After hideFindPopup
QuickFolders {QUICKMOVE} 9:53:37.679  [0 ms]	 
tearDownSearchBox()
QuickFolders {QUICKMOVE} 9:53:37.679  [0 ms]	 
findFolder(false, undefined)
QuickFolders 9:53:37.679  [0 ms]	 
toggleMoveModeSearchBox(false)
QuickFolders 9:53:44.621  [6942 ms]	 
persist tab category: undefined
QuickFolders 9:53:44.621  [0 ms]	 
persist tab category: undefined 2
QuickFolders 9:53:44.622  [1 ms]	 
persist tab category: undefined
```
Hope that's what you where looking for.

**Comment by RealRaven2000** [comment]
thanks for the debug log. Strangely, no errors are listed - I will have to try and reproduce the problem or add more detail to the debug log.

can you press F9 (if you use QuickFolders Pro) to make the folder tree visible? Or:
![image](https://user-images.githubusercontent.com/6314674/72430160-8a62cf00-3789-11ea-8207-348448403bad.png)

I need to see your current "Ordnerliste" for this. Or how do you get there - it appears the mails are in Inbox and 2020:
![image](https://user-images.githubusercontent.com/6314674/72430178-964e9100-3789-11ea-8e91-fdc157fc203c.png)

Again even with the "Themenbaum" you won't see the original mail unless you are in a special folder such as "gmail/all mail". Themenbaum will only give you what is visible within that folder; so if the emails are in different folders I don't understand how it can display that, unless you did a search. Search (and virtual folders) can display emails that are in various folders - which basically answers my own question on how to test this. I am still curious how you make it work.

**Comment by FrankMittelbach** [comment]
> thanks for the debug log. Strangely, no errors are listed - I will have to try and reproduce the problem or add more detail to the debug log.

sorry, guess I made myself not very clear then. It doesn't crash, ie no errors, it simply reports that nothing needs doing because the first message is in place. It gives me a nice popup saying so.

However, if the order is the other way around, ie the first message  is not already in the target folder it then it moves all messages than need moving.

**Comment by FrankMittelbach** [comment]
> Again even with the "Themenbaum" you won't see the original mail unless you are in a special folder such as "gmail/all mail". Themenbaum will only give you what is visible within that folder; so if the emails are in different folders I don't understand how it can display that, unless you did a search. Search (and virtual folders) can display emails that are in various folders - which basically answers my own question on how to test this. I am still curious how you make it work.

well, there must be some setting that makes that happen but don't ask me how I use this for ages, it is standard for me even across different accounts I just selected in my inbox a mail and Themenbaum did find me the same mail in the inbox of a different account. What it leaves out are mails I moved to the deleted folder but anything else is shown.

btw I use "Pro" but F9  doesn't work on the Mac version maybe it is assigned on a different key there,  anyway the mails are in Inbox and in archiv/2020 even if Ablageort only shows "2020"

**Comment by RealRaven2000** [comment]
> sorry, guess I made myself not very clear then. It doesn't crash, ie no errors, it simply reports that nothing needs doing because the first message is in place. It gives me a nice popup saying so.

ok, I can fix that. Still need to reproduce the conditions for it, so:

> well, there must be some setting that makes that happen but don't ask me how I use this for ages, it is standard for me even across different accounts I just selected in my inbox a mail and Themenbaum did find me the same mail **in the inbox** of a different account. 

ok, that's the key right there  - these are both inboxes. Sounds to me like you are using unified folders- is it set like this? Ansicht - Ordner - Gruppiert
![image](https://user-images.githubusercontent.com/6314674/72438981-ef282480-379d-11ea-8cf8-7ea183c157f4.png)

A screenshot of your folder tree and which node is selected would be super helpful. I thought you meant stuff in Sent (or another folder lower in the tree) plus your inbox. And (AFAIK) that's only possible with a virtual / search folder.

> What it leaves out are mails I moved to the deleted folder but anything else is shown.

but these are all Inbox folders, right? I believe "unified folder view" shows all inboxes in the same "posteingang". Another stupid question - do you use filters to move mails from one inbox to another inbox, or how do they end up in a different inbox? (I am always trying to empty my inbox and move mail to more topical filters, but YMMV)
 
> btw I use "Pro" but F9 doesn't work on the Mac version maybe it is assigned on a different key there, anyway the mails are in Inbox and in archiv/2020 even if Ablageort only shows "2020"

hmm that's strange, I checked the code and it's definitely there in the overlay:

```
<key id="quickFolders-ToggleTree" keycode="VK_F9" oncommand="QuickFolders.Interface.toggleFolderTree();"/>
```

**Comment by FrankMittelbach** [comment]
> ok, that's the key right there - these are both inboxes. Sounds to me like you are using unified folders- is it set like this? Ansicht - Ordner - Gruppiert
> ![image](https://user-images.githubusercontent.com/6314674/72438981-ef282480-379d-11ea-8cf8-7ea183c157f4.png)

sort of

<img width="793" alt="Screenshot 2020-01-15 at 15 09 12" src="https://user-images.githubusercontent.com/1062981/72441185-82ffed80-37aa-11ea-9b1a-6a5b25b5cee2.png">

> 
> A screenshot of your folder tree and which node is selected would be super helpful. I thought you meant stuff in Sent (or another folder lower in the tree) plus your inbox. And (AFAIK) that's only possible with a virtual / search folder.

just moved one of your replies into the Sent folder (which is called Gesendet) and using Themenbaum it finds it there, so I guess you are mistaken 

<img width="1421" alt="Screenshot 2020-01-15 at 15 15 47" src="https://user-images.githubusercontent.com/1062981/72441241-a32fac80-37aa-11ea-8cb5-5ecba7bc06b8.png">

> 
> > What it leaves out are mails I moved to the deleted folder but anything else is shown.
> 
> but these are all Inbox folders, right? I believe "unified folder view" shows all inboxes in the same "posteingang". Another stupid question - do you use filters to move mails from one inbox to another inbox, or how do they end up in a different inbox? (I am always trying to empty my inbox and move mail to more topical filters, but YMMV)

Not sure I have the right idea what "Inbox folders" are, on this Thunderbird I have afaik 2 inboxes from 2 different external accounts and a lot of topical boxes below one of them and one of the inboxes is rather huge.

> > btw I use "Pro" but F9 doesn't work on the Mac version maybe it is assigned on a different key there, anyway the mails are in Inbox and in archiv/2020 even if Ablageort only shows "2020"
> 
> hmm that's strange, I checked the code and it's definitely there in the overlay:

sorry, forget that one. the F9 works but only if not in a Themenbaum tab as that seems to be somehow the result of a search.

**Comment by RealRaven2000** [comment]
> > A screenshot of your folder tree and which node is selected would be super helpful. I thought you meant stuff in Sent (or another folder lower in the tree) plus your inbox. And (AFAIK) that's only possible with a virtual / search folder.
> 
> just moved one of your replies into the Sent folder (which is called Gesendet) and using Themenbaum it finds it there, so I guess you are mistaken

hmm, why does it work for you but not for me? Maybe it's because I use pop3; Imap isn't reliable enough. Also I do not use the Sent folder. :)

I need to do some testing...

> > 
> > hmm that's strange, I checked the code and it's definitely there in the overlay:
> 
> sorry, forget that one. the F9 works but only if not in a Themenbaum tab as that seems to be somehow the result of a search.

"if not in a Themenbaum tab" - I am confused :) there is a mode called "3-pane" window which is always the first tab in Thunderbird (it cannot be moved to a different position) - with that you can use F8 and F9 to toggle away folder tree (themenbaum) and preview (Vorschau). Then there are single message tabs (when you double-click an email). You also get a fresh "3-pane" tab if you open a folder in a new tab:

![image](https://user-images.githubusercontent.com/6314674/72443781-c0fb1280-37a6-11ea-8e0b-525d354f025b.png)

You seem to be using something else, "not 3 pane" tab with a list of emails - how do you get there? :)

**Comment by FrankMittelbach** [comment]
> > > A screenshot of your folder tree and which node is selected would be super helpful. I thought you meant stuff in Sent (or another folder lower in the tree) plus your inbox. And (AFAIK) that's only possible with a virtual / search folder.
> > 
> > 
> > just moved one of your replies into the Sent folder (which is called Gesendet) and using Themenbaum it finds it there, so I guess you are mistaken
> 
> hmm, why does it work for you but not for me? Maybe it's because I use pop3; Imap isn't reliable enough. Also I do not use the Sent folder. :)

well yes, I use imap usually, so that may be a difference. But I don't understand why.

The Sent folder was just an example because you mentioned that it should not work. 

> > sorry, forget that one. the F9 works but only if not in a Themenbaum tab as that seems to be somehow the result of a search.
> 
> "if not in a Themenbaum tab" - I am confused :) there is a mode called "3-pane" window which is always the first tab in Thunderbird (it cannot be moved to a different position) - with that you can use F8 and F9 to toggle away folder tree (themenbaum) and preview (Vorschau). 

yes right, that is what I meant. I mistakenly tried it first in the search folder opened by Themenbaum and there the F8/F9 have no effect. But when I use it normal tabs it does. That's way I said, please forget about the F9 it was just a misunderstanding where that works and where it doesn't

> You seem to be using something else, "not 3 pane" tab with a list of emails - how do you get there? :)

I use the 3 pane one and from there I get to the search ones by doing this:

<img width="786" alt="Screenshot 2020-01-15 at 16 21 36" src="https://user-images.githubusercontent.com/1062981/72446158-4c7aa080-37b3-11ea-9b2d-6d0c8c7afe46.png">

which then shows me a search tab even across accounts:

<img width="521" alt="Screenshot 2020-01-15 at 16 22 13" src="https://user-images.githubusercontent.com/1062981/72446268-86e43d80-37b3-11ea-97f1-b8c74994c9e5.png">

**Comment by RealRaven2000** [comment]
> > 
> > "if not in a Themenbaum tab" - I am confused :) there is a mode called "3-pane" window which is always the first tab in Thunderbird (it cannot be moved to a different position) - with that you can use F8 and F9 to toggle away folder tree (themenbaum) and preview (Vorschau).
> 
> yes right, that is what I meant. I mistakenly tried it first in the search folder opened by Themenbaum and there the F8/F9 have no effect. But when I use it normal tabs it does. That's way I said, please forget about the F9 it was just a misunderstanding where that works and where it doesn't
> 
> > You seem to be using something else, "not 3 pane" tab with a list of emails - how do you get there? :)
> 
> I use the 3 pane one and from there I get to the search ones by doing this:
> 
> <img alt="Screenshot 2020-01-15 at 16 21 36" width="786" src="https://user-images.githubusercontent.com/1062981/72446158-4c7aa080-37b3-11ea-9b2d-6d0c8c7afe46.png">
> 
> which then shows me a search tab even across accounts

AHHH. Now I get it. _Bei mir fällt der Groschen heute nur in Pfennigen_.
I think this is a search folder at heart. Good stuff this means I can finally start reproducing / testing the bug :) 
👍👍

**Comment by RealRaven2000** [comment]
Ok, I made a nice test case and a fix. One thing I was worried about was if there are mails in different folders (as I only pass one source folder to the copy service); but it seems that it works even if there are some messages in the target folder (no need to move them) most in the source folder and some in a third folder - all get moved as part of the same call with this version:

[QuickFolders-wx-5.2pre20.zip](https://github.com/RealRaven2000/QuickFolders/files/5842750/QuickFolders-wx-5.2pre20.zip)

Test
1) mails from a search:
![image](https://user-images.githubusercontent.com/6314674/105182589-1d64e080-5b25-11eb-9b61-b6fd986d54e9.png)

2) after dragging this selection to the folder "Egard" via the quickMove button:
![image](https://user-images.githubusercontent.com/6314674/105182630-2a81cf80-5b25-11eb-8eb1-beab6e689b96.png)

As you can see, the mails from both folder "dev" and "Inbox" were successfully moved to the new location.

**Comment by FrankMittelbach** [comment]
need to test a bit more, but looks good and so I guess this can be closed - thanks

**Comment by RealRaven2000** [comment]
PS: I usually close the issues when I release the new version (5.2). But it's ok. Thinking of backporting this for my Thunderbird 68 users, too.

**Comment by FrankMittelbach** [comment]
ok, so you do as we do in LaTeX. In that case I'm happy to leave it open ... process varies and many people do differently, but I'm also a fan to only close when it has been shipped as that seems easier for people to see if they have hit a problem which is still unresolved in the version they have on their machine.

**Comment by RealRaven2000** [comment]
Implemented in Releases 5.2 and 4.19.1 (published on 22/01/2021)
